### ISSUE TITLE
bug: Incorrect overload overlaps in sync, or missing overlap in async

### ISSUE THREAD

**Issue by mvanderlee** [microsoft/pyright#10759]
Continuation of discussion #10756

**Describe the bug**
I'm seeing `Overload 1 for "myfn" overlaps overload 2 and returns an incompatible type`
Eric says that's correct, and the bug is that the async function doesn't raise it. I disagree based on behaviour. 
The proposed solution was to not use any defaults but if that's truly the solution that overloads become a hinderance instead of useful, and I must stop using them in my project.

Expected behaviour:
I can define overloads to ensure that the types are accurate and the type checks and IDE hints such as VSCode's popup on hover, show me the function and the defaults. 
If I have a function with multiple kwargs, and remove the defaults from the overloads, then according to pyright the function has required args, not optional kwargs. This renders overloads useless? I may be missing something and am open to being educated. 

Our usecase is that based on some kwargs, a boolean in this case, the return type changes. And I'd like the type checker to be smart enough to detect that. async overloads work perfectly for us. 

**Code or Screenshots**
[Playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQAySMplZAsAFBcDGZlAZwFQAskiGo0AZWIoYqYmQBcXKGqgJBArlxEBJKVL0A5AOJQAvKPEDJMuQrIAKAJQ7OAYihS4KHlwABIlIKGi5qYmAoCDhgFCdVdUoQNCUoARgQWkS1AGsAd2S0AH0wFGI01HwrAEYABmzOdUwBYoyQSTSmFhA2AG0AFRAAV2IAXUsoIdHG5oKi4ph8sDTgUPwAHyhjMuIoLbEJdHt5crJJ-UMTU0aXKABaAD50zLSAOg-3QODyKmpwyLRWLxHJQIppdqzdTzFKlcqVOSTepQtTiNqZTqMZisMh9ABibAE40mBLIRJRUBhJSWKygayom22u321iO0lkp0UFwMRjMtwezyq70%2B3FFESiMTiCSaSRSEMyFKpcIqmERtQaoLR7UxACMwGBzlZSeTQUqaat1iyduUWYdbMcOY5uVc%2BVw7k8XiAWULQSBiDBhiAUFAAETAfUhzBRLUY9BQRREqAAFi%2BnEl8TDEfoMY66As02ILiAA)

[Playground without defaults in the overloads](https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQAySMplZAsAFBcDGZlAZwFQAskiGo0AZWIoYqYmQBcXKGqgJBArlxEBJKVL0A5AOJQAvKPEDJMuQrIAKAJQ7OAYihS4KHlwABIlIKGi5qYmAoCDhgFCdVdUoQNCUoARgQWkS1AGsAd2S0AH0wFGI01BhsznVMAWKMkEk0phYQNgBtABUQAFdiAF0auoKi4ph8sDTgUPwAHyhjMuIoRbEJdHt5crIalygAWgA%2BdMy0gDor90Dg8ipqcMjo2PicqCK0ppH1MZTS8qVOQ-NTiRqZFqMZisMidABibAEQxBUD%2BJUm0ygsyoCyWKzW1k20lkO0U%2ByOpyql2u3FpESiMTiCVqSRSX0yKLRAIqmDkligAEYAAwosFNSEAIzAYDI-IRZCRnMK-wxMzmBOW5QJG1sWxJjn5%2BkMJlM5JOZxABKp7xAxBgfRAKCgACJgNLnZgomKIegoIokVAACw3TiM%2BKu930b3NdAWXoDFxAA)

[Playground without defaults on the second overload](https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQAySMplZAsAFBcDGZlAZwFQAskiGo0AZWIoYqYmQBcXKGqgJBArlxEBJKVL0A5AOJQAvKPEDJMuQrIAKAJQ7OAYihS4KHlwABIlIKGi5qYmAoCDhgFCdVdUoQNCUoARgQWkS1AGsAd2S0AH0wFGI01HwrAEYABmzOdUwBYoyQSTSmFhA2AG0AFRAAV2IAXUsoIdHG5oKi4ph8sDTgUPwAHyhjMuIoLbEJdHt5crJJ-UMTU0aXKABaAD50zLSAOg-3QODyKmpwyLRWLxHJQIppdqzdTzFKlcqVORQtTiNqZTqMZisMh9ABibAE4yRUBhJSWKygayom22u321iO0lkp0Utwezyq70%2B3G5ESiMTiCSaSRSEMyRJJcIqmDkk3qRJR7XRACMwGBzlY8WQCeLCrCyat1nSduU6YdbMcmY4LgYjGZWU8XiA6RzQSBiDBhiAUFAAETAVU%2BzBRBVo9BQRQEqAAFi%2BnH58T9AfoIY66AsmoJLiAA) - Now setting the kwarg to False is no longer valid

[Playground with async](https://pyright-play.net/?pythonVersion=3.13&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQAySMplZAsAFBcDGZlAZwFQAskiGo0AZWIoYqYmQBcXKGqgJBArlxEBJKVL0A5AOJQAvKPEDJMuQrIAKAJQ7OAASKkKNLoLgUHihqYmAoCDhgFCdVdUoQNCUoARgQWji1AGsAdwS0AH0wFGJk1HwrAEYABgzOdUwBAtSQSWSmFhA2AG0AFRAAV2IAXUsofqG6htz8gpgcsGTgX3wAHyhjYuIodbEJdHt5ErIx-UMTUzqXKABaAD4UtOSAOlf3Li8ScipqfwFA4KhcKRaKxerxRLJFpTdQzRJFEplORjGowtTiZppNqMZisMjdABibAEIzGRLIJLRUDhhXmiygyyoaw2Wx21n20lkR0UpwMRjMV1uD3KLze3HFASCITCESiMUyUHyULSVJpCNKmGRVVqCoxLWxACMwGATlZyZSFWq6UsVmzNiU2XtbAcuY5eecBVxrvdHiA2SKFSBiDABiAUFAAETAY0RzDhPVY9BQRQkqAAFncksBMuIAA9iDwBixXCpwWpKHlmLLQVGY-QE610BYJsQ3JwgA) - This behaves exactly how I expect. 

```python
from typing import overload, Literal

class MissingSentinel:
    pass

MISSING = MissingSentinel()

# Sync
@overload
def myfn(
    arg: str,
    kwarg_one: int = 10,
    is_string: Literal[True] = True,
    kwarg_two: float | None | MissingSentinel = MISSING,
) -> str: ...


@overload
def myfn(
    arg: str,
    kwarg_one: int = 10,
    is_string: Literal[False] = False,
    kwarg_two: float | None | MissingSentinel = MISSING,
) -> int: ...


def myfn(
    arg: str,
    kwarg_one: int = 10,
    is_string: bool = False,
    kwarg_two: float | None | MissingSentinel = MISSING,
) -> str | int:
    return "foo" if is_string else 4


myfn("foo", is_string=True)
```
